src/EA.Iws.Domain/ImportNotificationAssessment/Transactions/ImportPaymentTransaction.cs

```diff
@@ -26,21 +27,39 @@ public class ImportPaymentTransaction
         public async Task Save(Guid notificationId, DateTime date, decimal amount, PaymentMethod paymentMethod,
             string receiptNumber, string comments)
         {
-            if (await transactionCalculator.PaymentIsNowFullyReceived(notificationId, amount))
+            var transaction = ImportNotificationTransaction.PaymentRecord(notificationId, date, amount,
+                paymentMethod, receiptNumber, comments);
+            var balance = await transactionCalculator.Balance(transaction.NotificationId)
+               - transaction.Credit.GetValueOrDefault()
+               + transaction.Debit.GetValueOrDefault();
+
+            transactionRepository.Add(transaction);
+
+            if (balance <= 0)
             {
                 var assessment = await assessmentRepository.GetByNotification(notificationId);
+                var transactions = await transactionRepository.GetTransactions(transaction.NotificationId);
+                transactions = transactions.Where(t => t.Credit > 0).OrderByDescending(t => t.Date).ToList();
 
-                if (assessment.Status == ImportNotificationStatus.AwaitingPayment)
+                foreach (var tran in transactions)
                 {
-                    assessment.PaymentComplete(date);
-                }
-                else if (!assessment.Dates.PaymentReceivedDate.HasValue)
-                {
-                    assessment.Dates.PaymentReceivedDate = date;
+                    if (balance == 0)
+                    {
+                        if (assessment.Status == ImportNotificationStatus.AwaitingPayment)
+                        {
+                            assessment.PaymentComplete(tran.Date);
+                        }
+                        else
+                        {
+                            assessment.Dates.PaymentReceivedDate = tran.Date;
+                        }
+
+                        break;
+                    }
+
+                    balance += tran.Credit.GetValueOrDefault() - tran.Debit.GetValueOrDefault();
```

Comment:
You are doing the same thing with almost the same code four times.  I think that it would be better to have a method in the transaction repository that sets the date and is called in these four places.

Reply:
Updated. Had to do it once for (export) notification and for import notification as these have 2 different repos.